File: generate_icon.py
#!/usr/bin/env python3
"""
Generate ASD Behavior Tracker app icon with rainbow infinity symbol on purple background
"""

from PIL import Image, ImageDraw
import math

def create_gradient_background(size, color1, color2):
    """Create a radial gradient background"""
    base = Image.new('RGB', (size, size), color1)
    draw = ImageDraw.Draw(base)

    # Create radial gradient
    center_x, center_y = size // 2, size // 2
    max_radius = math.sqrt(center_x**2 + center_y**2)

    for y in range(size):
        for x in range(size):
            # Calculate distance from center
            distance = math.sqrt((x - center_x)**2 + (y - center_y)**2)
            ratio = distance / max_radius

            # Interpolate colors
            r = int(color1[0] * (1 - ratio) + color2[0] * ratio)
            g = int(color1[1] * (1 - ratio) + color2[1] * ratio)
            b = int(color1[2] * (1 - ratio) + color2[2] * ratio)

            base.putpixel((x, y), (r, g, b))

    return base

def draw_infinity_symbol(draw, center_x, center_y, width, height, num_points=200):
    """
    Draw infinity symbol and return points for coloring
    The infinity symbol is a lemniscate: x = a*cos(t), y = a*sin(t)*cos(t)
    """
    points = []
    a = width / 2  # Scale factor
    b = height / 2

    for i in range(num_points):
        t = (2 * math.pi * i) / num_points

        # Lemniscate parametric equations
        denominator = 1 + math.sin(t)**2
        x = center_x + (a * math.cos(t)) / denominator
        y = center_y + (b * math.sin(t) * math.cos(t)) / denominator

        points.append((x, y))

    return points

def get_rainbow_color(position):
    """Get rainbow color based on position (0.0 to 1.0)"""
    # Rainbow: Red -> Orange -> Yellow -> Green -> Blue -> Indigo -> Violet
    colors = [
        (255, 0, 0),      # Red
        (255, 127, 0),    # Orange
        (255, 255, 0),    # Yellow
        (0, 255, 0),      # Green
        (0, 127, 255),    # Blue
        (75, 0, 130),     # Indigo
        (148, 0, 211),    # Violet
    ]

    # Determine which two colors to interpolate between
    scaled_pos = position * (len(colors) - 1)
    idx = int(scaled_pos)
    if idx >= len(colors) - 1:
        return colors[-1]

    # Interpolate between two adjacent colors
    ratio = scaled_pos - idx
    c1 = colors[idx]
    c2 = colors[idx + 1]

    r = int(c1[0] * (1 - ratio) + c2[0] * ratio)
    g = int(c1[1] * (1 - ratio) + c2[1] * ratio)
    b = int(c1[2] * (1 - ratio) + c2[2] * ratio)

    return (r, g, b)

def create_app_icon(output_path, size=1024):
    """Create the app icon"""
    # Purple gradient colors (lighter in center, darker on edges)
    color1 = (147, 112, 219)  # Medium Purple (center)
    color2 = (75, 0, 130)     # Indigo (edges)

    # Create base image with gradient
    img = create_gradient_background(size, color1, color2)
    draw = ImageDraw.Draw(img)

    # Calculate infinity symbol dimensions
    center_x = size // 2
    center_y = size // 2
    symbol_width = int(size * 0.65)
    symbol_height = int(size * 0.3)

    # Draw infinity symbol with rainbow colors
    points = draw_infinity_symbol(draw, center_x, center_y, symbol_width, symbol_height, 400)

    # Draw the infinity symbol with thick rainbow-colored lines
    line_width = int(size * 0.08)  # Thick line

    for i in range(len(points) - 1):
        # Get rainbow color based on position along the curve
        position = i / len(points)
        color = get_rainbow_color(position)

        # Draw thick line segment
        x1, y1 = points[i]
        x2, y2 = points[i + 1]

        # Draw multiple lines to create thickness
        for offset in range(-line_width//2, line_width//2):
            draw.line(
                [(x1, y1 + offset), (x2, y2 + offset)],
                fill=color,
                width=3
            )
            draw.line(
                [(x1 + offset, y1), (x2 + offset, y2)],
                fill=color,
                width=3
            )

    # Add rounded corners for iOS app icon style
    mask = Image.new('L', (size, size), 0)
    mask_draw = ImageDraw.Draw(mask)

    # iOS icon corner radius is approximately 22.37% of the icon size
    corner_radius = int(size * 0.2237)
    mask_draw.rounded_rectangle(
        [(0, 0), (size, size)],
        radius=corner_radius,
        fill=255
    )

    # Apply rounded corners
    output = Image.new('RGBA', (size, size), (0, 0, 0, 0))
    output.paste(img, (0, 0))
    output.putalpha(mask)

    # Convert back to RGB for PNG
    final = Image.new('RGB', (size, size), (255, 255, 255))
    final.paste(output, (0, 0), output)

    # Save the icon
    final.save(output_path, 'PNG', quality=100)
    print(f"Icon saved to: {output_path}")

if __name__ == "__main__":
    output_file = "BehaviorTracker/Assets.xcassets/AppIcon.appiconset/asd-icon-1024.png"
    create_app_icon(output_file, 1024)
    print("\nApp icon created successfully!")
    print(f"Size: 1024x1024")
    print("Features:")
    print("  - Purple gradient background (lighter to darker)")
    print("  - Rainbow infinity symbol (neurodiversity/ASD symbol)")
    print("  - Rounded corners for iOS style")
    print("\nNext steps:")
    print("  1. Update Contents.json to use 'asd-icon-1024.png'")
    print("  2. Delete old 'patterns-1024.png'")
    print("  3. Build and run the app to see the new icon")

File: generate_icon.swift
#!/usr/bin/env swift

import Foundation
import AppKit
import CoreGraphics

// MARK: - Helper Functions

func createGradientBackground(size: CGFloat) -> NSImage {
    let image = NSImage(size: NSSize(width: size, height: size))
    image.lockFocus()

    let context = NSGraphicsContext.current!.cgContext

    // Purple gradient colors
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    let colors = [
        CGColor(red: 147/255, green: 112/255, blue: 219/255, alpha: 1.0), // Medium Purple (center)
        CGColor(red: 75/255, green: 0/255, blue: 130/255, alpha: 1.0)      // Indigo (edges)
    ]

    let gradient = CGGradient(
        colorsSpace: colorSpace,
        colors: colors as CFArray,
        locations: [0.0, 1.0]
    )!

    // Draw radial gradient
    let center = CGPoint(x: size/2, y: size/2)
    let radius = size * 0.7

    context.drawRadialGradient(
        gradient,
        startCenter: center,
        startRadius: 0,
        endCenter: center,
        endRadius: radius,
        options: [.drawsBeforeStartLocation, .drawsAfterEndLocation]
    )

    image.unlockFocus()
    return image
}

func getRainbowColor(position: CGFloat) -> NSColor {
    // Rainbow colors array
    let colors: [(r: CGFloat, g: CGFloat, b: CGFloat)] = [
        (255, 0, 0),      // Red
        (255, 127, 0),    // Orange
        (255, 255, 0),    // Yellow
        (0, 255, 0),      // Green
        (0, 127, 255),    // Blue
        (75, 0, 130),     // Indigo
        (148, 0, 211),    // Violet
    ]

    let scaledPos = position * CGFloat(colors.count - 1)
    let idx = Int(scaledPos)

    guard idx < colors.count - 1 else {
        let last = colors.last!
        return NSColor(red: last.r/255, green: last.g/255, blue: last.b/255, alpha: 1.0)
    }

    let ratio = scaledPos - CGFloat(idx)
    let c1 = colors[idx]
    let c2 = colors[idx + 1]

    let r = (c1.r * (1 - ratio) + c2.r * ratio) / 255
    let g = (c1.g * (1 - ratio) + c2.g * ratio) / 255
    let b = (c1.b * (1 - ratio) + c2.b * ratio) / 255

    return NSColor(red: r, green: g, blue: b, alpha: 1.0)
}

func drawInfinitySymbol(context: CGContext, centerX: CGFloat, centerY: CGFloat, width: CGFloat, height: CGFloat, lineWidth: CGFloat) {
    let numPoints = 500
    let a = width / 2
    let b = height / 2

    // Generate points for infinity symbol (lemniscate)
    for i in 0..<numPoints {
        let t = (2 * CGFloat.pi * CGFloat(i)) / CGFloat(numPoints)
        let position = CGFloat(i) / CGFloat(numPoints)

        // Lemniscate parametric equations
        let denominator = 1 + sin(t) * sin(t)
        let x = centerX + (a * cos(t)) / denominator
        let y = centerY + (b * sin(t) * cos(t)) / denominator

        // Get rainbow color for this position
        let color = getRainbowColor(position: position)

        // Draw thick point
        context.setFillColor(color.cgColor)
        context.fillEllipse(in: CGRect(x: x - lineWidth/2, y: y - lineWidth/2, width: lineWidth, height: lineWidth))
    }
}

func createAppIcon(size: CGFloat, outputPath: String) {
    // Create background
    let background = createGradientBackground(size: size)

    // Create final image
    let finalImage = NSImage(size: NSSize(width: size, height: size))
    finalImage.lockFocus()

    // Draw background
    background.draw(at: .zero, from: NSRect(x: 0, y: 0, width: size, height: size), operation: .copy, fraction: 1.0)

    // Get graphics context
    guard let context = NSGraphicsContext.current?.cgContext else {
        print("Failed to get graphics context")
        return
    }

    // Draw infinity symbol
    let symbolWidth = size * 0.65
    let symbolHeight = size * 0.3
    let lineWidth = size * 0.08

    drawInfinitySymbol(
        context: context,
        centerX: size / 2,
        centerY: size / 2,
        width: symbolWidth,
        height: symbolHeight,
        lineWidth: lineWidth
    )

    finalImage.unlockFocus()

    // Save as PNG with exact size
    guard let tiffData = finalImage.tiffRepresentation,
          let bitmapImage = NSBitmapImageRep(data: tiffData) else {
        print("Failed to create bitmap data")
        return
    }

    // Force exact pixel size
    bitmapImage.size = NSSize(width: size, height: size)

    guard let pngData = bitmapImage.representation(using: .png, properties: [:]) else {
        print("Failed to create PNG data")
        return
    }

    do {
        try pngData.write(to: URL(fileURLWithPath: outputPath))
        print("Icon saved to: \(outputPath)")
        print("Size: \(Int(size))x\(Int(size))")
        print("Features: Purple gradient background with rainbow infinity symbol")
    } catch {
        print("Error saving icon: \(error)")
    }
}

// MARK: - Main

let outputPath = "BehaviorTracker/Assets.xcassets/AppIcon.appiconset/asd-icon-1024.png"

// Remove old icon if it exists
let fileManager = FileManager.default
if fileManager.fileExists(atPath: outputPath) {
    try? fileManager.removeItem(atPath: outputPath)
}

createAppIcon(size: 1024, outputPath: outputPath)

print("\nApp icon created successfully!")
print("\nNext steps:")
print("  1. Update Contents.json to use 'asd-icon-1024.png'")
print("  2. Delete old 'patterns-1024.png'")
print("  3. Build and run the app to see the new icon")
